### DISCUSSION TITLE
Nodes performance

### DISCUSSION THREAD

**Discussion by jonesa99** [kubernetes-sigs/kwok#787] [Q&A]
Hi, I'm starting to look into kwok and from a blog I read that "Kwok doesn’t accurately reflect the performance of real nodes because it approximates behaviors using simple formulas". Is it possible to find these formulas somewhere? And how can a fake node simulate a job? Do you need to also run some real nodes? Thanks for the help

**Comment by wzshiming** [comment]
Refer to https://kwok.sigs.k8s.io/docs/user/stages-configuration/

**Reply by jonesa99** [reply]
I'm sorry but I don't see any formula there

**Reply by wzshiming** [reply]
The Stage API is the formula, which describes how the resource changes.

**Reply by jonesa99** [reply]
maybe my question was not clear. I don't understand after creating how a fake resource can simulate a job. Does it need any data from real resources?

**Reply by wzshiming** [reply]
On the control plane it's all just data, kwok just simulates the data, it doesn't need to be anything real.

**Reply by jonesa99** [reply]
reading the discussions there it seems possible for example to run an argo workflow in kwok. in that case, how can the fake resources simulate a workflow?

**Reply by wzshiming** [reply]
Refer to https://kwok.sigs.k8s.io/docs/examples/argo/

**Reply by jonesa99** [reply]
yeah this is the one I was referring to. In that case, how can I extract the metrics of a pipeline if the resources are fake?

**Reply by wzshiming** [reply]
I'm not familiar with Argo, but in the future kwok could consider supporting the simulation of these metrics if it is known what metrics can be extracted.

my plan for the next release is to improve the simulation of metrics, currently only metrics-server, custom.metrics, and external.metrics of k8s are planned.

**Reply by jonesa99** [reply]
okay thanks. for example, now if I simulate the resources in kwok have I the access to metrics like time required for completation or CPU usage?

**Reply by wzshiming** [reply]
Sure, the [simulation of CPU/Memory utilization](https://github.com/kubernetes-sigs/kwok/pull/740/files) is complete, but it will go to release 0.5.

**Reply by jonesa99** [reply]
superb! in easy terms, how is the time required for a workflow computed if the resources are only fakes? I thought it was based on the usage of both real and fake ones but I'm wrong according to your link

**Reply by wzshiming** [reply]
These can be simulated at will, the main problem is how to provide the data.

**Reply by wzshiming** [reply]
This project is actually very simple, the main thing is what data the controller needs, the project to provide this data, as well as provide an easy way to configure this data.

**Reply by jonesa99** [reply]
in the case of Argo it requires the input file and the steps of the pipeline, is this the data that you are referring to?

**Reply by wzshiming** [reply]
Yes, that will be included in the future, but right now it's mostly about nodes, pods, and metrics.

**Reply by jonesa99** [reply]
okay, and how can you estimate metrics like time if you don't know the data ut just the steps?

**Reply by wzshiming** [reply]
I think that can be from a label or field, or just a constant.

**Reply by jonesa99** [reply]
So I don't get what kwok is simulating. I can create different nodes and pods, assign a pipeline to them and get metrics like time and cpu. How are these metrics generated? And what are they useful for? In my mind it seemed to be useful to simulate the run of a pipeline even if my resources are not enough due to big data volume

**Reply by wzshiming** [reply]
I don't know how Argo collects this data, but there is always a way to simulate it, give Argo the data it needs, kwok doesn't care what it is.

https://kwok.sigs.k8s.io/docs/design/introduction/#user-stories

In fact, the main goal is unit testing and large-scale testing.

**Reply by wzshiming** [reply]
> How are these metrics generated? And what are they useful for?

I can't answer this question at the moment, because I'm not sure how Argo works.

**Reply by jshin48** [reply]
> Yes, that will be included in the future, but right now it's mostly about nodes, pods, and metrics.

Hello, is this feature applied in the current version?
I am also wondering how to put the execution time of each pod for the simulation. 
Does it have to be specified within a yaml file when pod creation request is submitted?
If so, which field? Is there any example yaml file that contains this information?

**Reply by wzshiming** [reply]
> > Yes, that will be included in the future, but right now it's mostly about nodes, pods, and metrics.
> 
> Hello, is this feature applied in the current version? I am also wondering how to put the execution time of each pod for the simulation. Does it have to be specified within a yaml file when pod creation request is submitted? If so, which field? Is there any example yaml file that contains this information?

It feels like your question is out of context with this discussion... 

If you want to change the execution time of a pod, you can to changing the default stages configuration. [Here is a basic principle and 2 basic cases](https://kwok.sigs.k8s.io/docs/user/stages-configuration/#pod-stages-that-simulate-real-behavior-as-closely-as-possible), the main action is to configure the delay property of the pod stage.  More examples can be found in [the github search](https://github.com/search?q=%22apiVersion%3A+kwok.x-k8s.io%2Fv1alpha1%22+%22delay%3A%22+language%3AYAML+-is%3Afork&type=code) on how to used by other user

**Comment by jonesa99** [comment]
okay so in general it can be used to simulate cluster in large scale, but also in this case how could you simulate to have huge amount of nodes? how do the fake resources behave?

**Reply by wzshiming** [reply]
> how could you simulate to have huge amount of nodes?

https://kwok.sigs.k8s.io/docs/user/kwok-manage-nodes-and-pods/#update-spec-of-nodes-or-pods

> how do the fake resources behave?

https://kwok.sigs.k8s.io/docs/user/stages-configuration/

**Reply by jonesa99** [reply]
I have already checked it. The fake resources are API objects, but how can they simulate real ones? Honestly I don't find it very clear.. I would like to contribute and write a brief description in easier terms, but I'm still studying it

**Reply by wzshiming** [reply]
Simulate real ones? this is a very large amount of work, we need to describe the behavior of each change in the resource and provide a way to configure it, this may be possible to try to do it after the main function is complete.

https://github.com/kubernetes-sigs/kwok/blob/main/kustomize/stage/pod/general/README.md

**Reply by jonesa99** [reply]
okay.so atm kwok is not replicating real behavior, what is the main function then?

**Reply by jonesa99** [reply]
and also, when I make a simulation using Argo and 5 or 10 fake resources, what can I learn from it?

**Reply by wzshiming** [reply]
> what is the main function then?

https://kwok.sigs.k8s.io/docs/design/introduction/#user-stories

> when I make a simulation using Argo and 5 or 10 fake resources, what can I learn from it?

Learned to use tools 😄

**Reply by jonesa99** [reply]
yeah but I mean, if it's not clear how the resources fake the process is not clear the difference between using 5 or 10 of them

**Reply by wzshiming** [reply]
I feel like I'm not understanding you correctly, I don't understand what you're trying to do, can you explain it step by step?

**Reply by jonesa99** [reply]
yeah. I found the guide for using kwok and argo together but it's not clear to me what is the main difference betweeen using just argo or using argo with kwok. I understand that the resources in the second case would be fake but I don't understand how it runs then

**Reply by wzshiming** [reply]
It's no different for argo, it still treats the resource as real, kwok just fakes the resource to trick argo according to argo's logic, you may need to look at the argo controller's logic to determine the details.

**Reply by jonesa99** [reply]
do you have any resources useful for that?

**Reply by wzshiming** [reply]
No, I didn't actually use Argo, that example was just me observing Argo's behavior and then writing a sample document that ran to prove that kwok could be used to test Argo, and hoping that someone would help me refine it.

**Reply by jonesa99** [reply]
okay good. and if i want to follow the current guide, how can i simulate using 10 or 20 nodes/pods?

**Reply by wzshiming** [reply]
Create the resources you want directly https://kwok.sigs.k8s.io/docs/user/kwok-manage-nodes-and-pods/

**Reply by jonesa99** [reply]
Okay, I have some issues since the heelo-world pipeline is running for over 10 minutes, I'm just following the guide so it's a bit unclear to me

**Reply by wzshiming** [reply]
You may have missed the first step.

https://kwok.sigs.k8s.io/docs/examples/argo/#custom-pod-behavior

**Reply by jonesa99** [reply]
isn't it just the renaming of the downloaded file?

**Reply by wzshiming** [reply]
``` bash
sed 's/Job/Workflow/g' pod-fast.yaml > workflow-fast.yaml
```

Replace Job with Workflow

**Reply by jonesa99** [reply]
I found the issue. Now the hello-world.yaml works fine, but when I use a more complex pipeline with artifacts I have an error. For reference, you can replicate the error using https://github.com/argoproj/argo-workflows/blob/master/examples/artifact-passing.yaml

**Reply by jonesa99** [reply]
I solved it using the optional statement

**Reply by jonesa99** [reply]
now I can work with complex pipelines, the only problem is that I get 0s for each step even if it takes some seconds. Which is the best way to collect metrics using kwok? I mean, to see standard resource consumption like cpu or time. Maybe prometheus?

**Reply by wzshiming** [reply]
https://kwok.sigs.k8s.io/docs/user/stages-configuration/#pod-stages-that-simulate-real-behavior-as-closely-as-possible

Pod behavior that you can change to suit your needs

> Which is the best way to collect metrics using kwok? I mean, to see standard resource consumption like cpu or time. Maybe prometheus?

I am not sure argo is how to collect this, but the [ResourceUsage API](https://github.com/kubernetes-sigs/kwok/pull/740 ) is WIP, it will simulate the resource usage (CPU/memory) according to k8s standard.

**Reply by jonesa99** [reply]
Okay thanks. But so without the additional API for metrics what could I get from kwok, expect for total time?

**Reply by wzshiming** [reply]
I am not sure argo is how to collect this

**Reply by jonesa99** [reply]
So what that's the advantage of using Argo with kwok?

**Reply by jonesa99** [reply]
I just read here https://kwok.sigs.k8s.io/docs/design/architecture/ that prometheus is available to collect metrics, do you have a guide for integrating it?

**Reply by jonesa99** [reply]
found it

**Reply by jonesa99** [reply]
sorry for not updating you, I made some progress and kwok seems useful for my case. I just have a quick question; in order to scale a pipeline using Argo I should expect the CPU and memory to increase when moving from 1 to 2 pods, while the execution time should decrease. Is it correct?  And should I see this behavior also in kwok?

**Reply by wzshiming** [reply]
Yes, Simulation of CPU/Memory usage is expected to be supported in the next release.

**Reply by jonesa99** [reply]
okay thanks. so atm in kwok I can only see the difference in time execution when changing the number of pods, right?

**Reply by wzshiming** [reply]
I think the time you can get it directly from the Pod

**Comment by jonesa99** [comment]
yes I can do it, I am just wondering what are the possible metrics to collect

**Reply by wzshiming** [reply]
The metric also provides the execution time, but I don't know which one argo uses, the one from the Pod or the one from the metric.
